### PR TITLE
[wip] LSAN integration

### PR DESCRIPTION
**Pull request by iliabylich**
Spoiler: this is a WIP project.

By enabling `-O0 -fsanitize=address` and running `./miniruby -v` with `ASAN_OPTIONS=detect_leaks=1:malloc_context_size=100 LSAN_OPTIONS=report_objects=1` I was able to identify 2 memory leaks in prelude. Not even in any real-world complex Ruby code.

1. `alias foo foo` - this code prevents method `foo` from GC-ing. Very artificial, I guess that's why nobody noticed it. However, defining a method and aliasing it on itself in a loop eats memory indefinitely. 1 leak per one method-aliased-on-itself, caught in prelude/kernel that does `alias pp pp`.
2. Method call with kwargs leaves garbage after GC-ing. Basically when `compile_call` is called it allocates `rb_callinfo_kwarg` inside `rb_callinfo`, but the wrapper is an `imemo` that is fully managed by GC. This `imemo` had a very basic cleanup mechanism that didn't `free` any heap-allocated fields. 1 leak per one **method call definition**, not per actual call, so it's also minor.

I also added some dumb code that cleans the stack and all Ruby heap objects. This PR is WIP and so for now I'm going to keep it there (of course it will be hidden behind some feature flag, TBD).

Current state: `./miniruby -e 'p 42'` shows 0 leaks (well, because there's almost no Ruby code). `make` doesn't complete, it aborts on generating `transdb.h` via `tool/generic_erb.rb`. The error comes from ASAN, not LSAN. Initially there was a `use-after-poison` error from `rb_str_new`, and so I blindly added `asan_unpoison_memory_region(RSTRING_PTR(str), len + termlen, true);` to `str_new0`, but I'm pretty sure it's wrong. @peterzhu2118 Is it related to RVARGC changes?

Anyway, I added it locally and the error has gone. Then other appeared (`unknown-crash on address ...`):

``` sh
./miniruby -I./lib -I. -I.ext/common ./tool/generic_erb.rb -c -o transdb.h ./template/transdb.h.tmpl ./enc/trans enc/trans
==14131==ERROR: AddressSanitizer: unknown-crash on address 0x00010b1c0cd5 at pc 0x000102885d20 bp 0x7ffeed93a2f0 sp 0x7ffeed93a2e8
READ of size 8 at 0x00010b1c0cd5 thread T0
    #0 0x105e65b14 in __asan_memset (/usr/local/opt/llvm/lib/clang/13.0.0/lib/darwin/libclang_rt.asan_osx_dynamic.dylib:x86_64+0x42b14)
    #1 0x105581925 in rb_str_format ruby/sprintf.c:262:5
    #2 0x1055d7b47 in rb_str_format_m ruby/string.c:2380:16
    #3 0x1057733ac in ractor_safe_call_cfunc_1 ruby/./vm_insnhelper.c:2861:12
    #4 0x1057525f5 in vm_call_cfunc_with_frame ruby/./vm_insnhelper.c:3037:11
    #5 0x10575ed3c in vm_sendish ruby/./vm_insnhelper.c:4751:15
```

The worst part about it is that I'm getting the same error on the latest master. Any suggestions? I feel like there must be a `poison` call somewhere in the allocation mechanism, not in individual `String` constructors.

It can also can be reproduced with

```
./miniruby -e 'open("./enc/trans/big5.c") { |f| f.each_line { |l| p l } }'
```

it stops at line 9222 with `#define to_Big5_HKSCS_E4_AE_offsets 49372`, absolutely no idea why, it's the same as 20% of the file.

### CODE REVIEW

**Comment by peterzhu2118** [comment]
Thank you for this patch. Getting ASAN/LSAN properly supported with integration on CI will be a very beneficial feature.

I'll spend more time sometime this week or next week to investigate the ASAN crashes. I wouldn't be surprised if Variable Width Allocation (RVARGC) was the culprit behind it.

If you could open separate PRs with small reproduction scripts for the memory leaks you've found, I'd be more than happy to take a look at it and merge it separately. If it affects released versions of Ruby (2.7, 3.0, 3.1), then could you also please open a ticket on the bug tracker? That way we can get a backport to those versions. [Here's an example of a ticket I've opened for a memory leak](https://bugs.ruby-lang.org/issues/18264).

**Comment by iliabylich** [comment]
> I'll spend more time sometime this week or next week to investigate the ASAN crashes

Thanks!

> If you could open separate PRs with small reproduction scripts for the memory leaks you've found

Sure, will do tomorrow.

**Comment by nobu** [comment]
Aliasing to the same name is intentionally adding the count, to suppress the later method redefinition warning.

File: string.c [hunk 1]
```diff
@@ -899,6 +899,7 @@ str_new0(VALUE klass, const char *ptr, long len, int termlen)
             rb_xmalloc_mul_add_mul(sizeof(char), len, sizeof(char), termlen);
 	STR_SET_NOEMBED(str);
     }
+    asan_unpoison_memory_region(RSTRING_PTR(str), len + termlen, true);
```
Comment (nobu) [inline string.c:902]: I think this is no longer needed by 16e7585557c338563447861ee50f835514bb7e3c.

**Pull request closed without merging**